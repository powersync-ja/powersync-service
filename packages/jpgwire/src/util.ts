import * as tls from 'tls';
import { DEFAULT_CERTS } from './certs.js';
import * as pgwire from './pgwire.js';
import { PgType } from './pgwire_types.js';
import * as datefns from 'date-fns';

// TODO this is duplicated, but maybe that is ok
export interface NormalizedConnectionConfig {
  id: string;
  tag: string;

  hostname: string;
  port: number;
  database: string;

  username: string;
  password: string;

  sslmode: 'verify-full' | 'verify-ca' | 'disable';
  cacert: string | undefined;

  client_certificate: string | undefined;
  client_private_key: string | undefined;
}

type Mutable<T> = {
  -readonly [P in keyof T]: T[P];
};

export interface PgWireConnectionOptions extends NormalizedConnectionConfig {
  resolved_ip?: string;
}

export function clientTlsOptions(options: PgWireConnectionOptions): tls.ConnectionOptions {
  if (options.client_certificate && options.client_private_key) {
    return {
      cert: options.client_certificate,
      key: options.client_private_key
    };
  } else {
    return {};
  }
}
export function tlsOptions(options: PgWireConnectionOptions): false | tls.ConnectionOptions {
  if (options.sslmode == 'disable') {
    return false;
  } else if (options.sslmode == 'verify-full') {
    // Validate against default CA, or supplied CA.
    // Also validate hostname.
    return {
      ca: options.cacert ?? DEFAULT_CERTS,
      // hostname for TLS validation.
      // This may be different from the host we're connecting to if we pre-resolved
      // the IP.
      host: options.hostname,
      servername: options.hostname,
      ...clientTlsOptions(options)
    };
  } else if (options.sslmode == 'verify-ca') {
    if (options.cacert == null) {
      throw new Error(`cacert required for verify-ca`);
    }
    return {
      ca: options.cacert!,
      host: options.hostname,
      servername: options.hostname,
      // Disable hostname check
      checkServerIdentity: () => undefined,
      ...clientTlsOptions(options)
    };
  } else {
    throw new Error(`Unsupported sslmode: ${options.sslmode}`);
  }
}

export async function connectPgWire(config: PgWireConnectionOptions, options?: { type?: 'standard' | 'replication' }) {
  let connectionOptions: Mutable<ExtendedPgwireOptions> = {
    application_name: 'PowerSync',

    // tlsOptions below contains the original hostname
    hostname: config.resolved_ip ?? config.hostname,
    port: config.port,
    database: config.database,

    user: config.username,
    password: config.password
  };

  if (options?.type != 'standard') {
    connectionOptions.replication = 'database';
  }

  if (config.sslmode != 'disable') {
    connectionOptions.sslmode = 'require';

    // HACK: Not standard pgwire options
    // Just the easiest way to pass on our config to pgwire_node.js
    connectionOptions.sslrootcert = tlsOptions(config);
  } else {
    connectionOptions.sslmode = 'disable';
  }
  const connection = await pgwire.pgconnect(connectionOptions as pgwire.PgConnectOptions);
  // HACK: Replace row decoding with our own implementation
  (connection as any)._recvDataRow = _recvDataRow;
  return connection;
}

function _recvDataRow(this: any, _message: any, row: Uint8Array[], batch: any) {
  for (let i = 0; i < this._rowColumns.length; i++) {
    const valbuf = row[i];
    if (valbuf == null) {
      continue;
    }
    const { binary, typeOid } = this._rowColumns[i];
    // TODO avoid this._clientTextDecoder.decode for bytea
    row[i] = binary
      ? // do not valbuf.slice() because nodejs Buffer .slice does not copy
        // TODO but we not going to receive Buffer here ?
        Uint8Array.prototype.slice.call(valbuf)
      : PgType.decode(this._rowTextDecoder.decode(valbuf), typeOid);
  }
  batch.rows.push(row);
}

export interface PgPoolOptions {
  /**
   * Maximum number of open connections.
   *
   * Once this limit is reached, queries are queued.
   *
   * Defaults to 5.
   */
  maxSize?: number;

  /**
   * Idle timeout in ms before a connection is closed.
   */
  idleTimeout?: number | undefined;
}

/**
 * Open a postgres pool.
 *
 * The pool cannot be used for transactions, and cannot be used for logical replication.
 */
export function connectPgWirePool(config: PgWireConnectionOptions, options?: PgPoolOptions) {
  const idleTimeout = options?.idleTimeout;
  const maxSize = options?.maxSize ?? 5;

  let connectionOptions: Mutable<ExtendedPgwireOptions> = {
    application_name: 'PowerSync',

    // tlsOptions below contains the original hostname
    hostname: config.resolved_ip ?? config.hostname,
    port: config.port,
    database: config.database,

    user: config.username,
    password: config.password,

    _poolSize: maxSize,
    _poolIdleTimeout: idleTimeout
  };

  if (config.sslmode != 'disable') {
    connectionOptions.sslmode = 'require';

    // HACK: Not standard pgwire options
    // Just the easiest way to pass on our config to pgwire_node.js
    connectionOptions.sslrootcert = tlsOptions(config);
  } else {
    connectionOptions.sslmode = 'disable';
  }

  const pool = pgwire.pgpool(connectionOptions as pgwire.PgConnectOptions);
  const originalGetConnection = (pool as any)._getConnection;
  (pool as any)._getConnection = function (this: any) {
    const con = originalGetConnection.call(this);
    // HACK: Replace row decoding with our own implementation
    (con as any)._recvDataRow = _recvDataRow;
    return con;
  };
  return pool;
}

/**
 * Hack: sslrootcert is passed through as-is to pgwire_node.
 *
 * We use that to pass in custom TLS options, without having to modify pgwire itself.
 */
export interface ExtendedPgwireOptions extends pgwire.PgConnectKnownOptions {
  sslrootcert?: false | tls.ConnectionOptions;
}

export function lsnMakeComparable(text: string) {
  const [h, l] = text.split('/');
  return h.padStart(8, '0') + '/' + l.padStart(8, '0');
}

/**
 * Convert a postgres timestamptz to a SQLite-compatible/normalized timestamp.
 *
 * We output in the format:
 * YYYY-MM-DD HH:MM:SSZ
 * YYYY-MM-DD HH:MM:SS.SSSZ
 *
 * Precision is preserved from the source string, with the exception of ".000".
 *
 * We have specific exceptions for -infinity and infinity.
 */
export function timestamptzToSqlite(source?: string) {
  if (source == null) {
    return null;
  }
  // Make compatible with SQLite
  const match = /^([\d\-]+) ([\d:]+)(\.\d+)?([+-][\d:]+)$/.exec(source);
  if (match == null) {
    if (source == 'infinity') {
      return '9999-12-31 23:59:59Z';
    } else if (source == '-infinity') {
      return '0000-01-01 00:00:00Z';
    } else {
      return null;
    }
  }
  const [_, date, time, precision, timezone] = match as any;
  // datefns.parseISO() supports some dates that new Date() doesn't, e.g. '0022-09-08 21:19:28+00'
  const parsed = datefns.parseISO(`${date}T${time}${timezone}`);
  // Some dates are outside the range of JS and/or SQLite, e.g. 10022-09-08 21:19:28+00.
  // Just fallback to null for now.
  // We may consider preserving the source value in these cases.
  if (isNaN(parsed.getTime())) {
    return null;
  }
  const text = parsed.toISOString().replace('T', ' ').replace('.000', '').replace('Z', '');

  return `${text}${precision ?? ''}Z`;
}

/**
 * For timestamp without timezone, we keep it mostly as-is.
 *
 * We make specific exceptions for 'infinity' and '-infinity'.
 *
 * https://www.postgresql.org/docs/current/datatype-datetime.html#DATATYPE-DATETIME-SPECIAL-VALUES
 */
export function timestampToSqlite(source?: string) {
  if (source == null) {
    return null;
  }
  if (source == 'infinity') {
    return '9999-12-31 23:59:59';
  } else if (source == '-infinity') {
    return '0000-01-01 00:00:00';
  } else {
    return source;
  }
}
/**
 * For date, we keep it mostly as-is.
 *
 * We make specific exceptions for 'infinity' and '-infinity'.
 *
 * https://www.postgresql.org/docs/current/datatype-datetime.html#DATATYPE-DATETIME-SPECIAL-VALUES
 */
export function dateToSqlite(source?: string) {
  if (source == null) {
    return null;
  }
  if (source == 'infinity') {
    return '9999-12-31';
  } else if (source == '-infinity') {
    return '0000-01-01';
  } else {
    return source;
  }
}

/**
 * PgResult rows are arrays of values.
 *
 * This converts it to objects.
 */
export function pgwireRows(rs: pgwire.PgResult): Record<string, any>[] {
  const columns = rs.columns;
  return rs.rows.map((row) => {
    let r: Record<string, any> = {};
    for (let i = 0; i < columns.length; i++) {
      const c = columns[i];
      r[c.name] = row[i];
    }
    return r;
  });
}
