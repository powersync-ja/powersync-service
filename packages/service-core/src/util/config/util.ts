import * as urijs from 'uri-js';

export function getSupabaseJwksUrl(connection: any) {
  if (connection == null) {
    return null;
  } else if (connection.type != 'postgresql') {
    return null;
  }

  let hostname: string | undefined = connection.hostname;
  if (hostname == null && typeof connection.uri == 'string') {
    hostname = urijs.parse(connection.uri).host;
  }
  if (hostname == null) {
    return null;
  }

  const match = /db.(\w+).supabase.co/.exec(hostname);
  if (match == null) {
    return null;
  }
  const projectId = match[1];

  return `https://${projectId}.supabase.co/auth/v1/.well-known/jwks.json`;
}
