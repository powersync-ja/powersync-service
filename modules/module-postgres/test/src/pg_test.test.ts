import * as pgwire from '@powersync/service-jpgwire';
import {
  applyRowContext,
  CompatibilityContext,
  SqliteInputRow,
  DateTimeValue,
  TimeValue,
  CompatibilityEdition
} from '@powersync/service-sync-rules';
import { describe, expect, test } from 'vitest';
import { clearTestDb, connectPgPool, connectPgWire, TEST_URI } from './util.js';
import { WalStream } from '@module/replication/WalStream.js';
import { PostgresTypeCache } from '@module/types/custom.js';

describe('pg data types', () => {
  async function setupTable(db: pgwire.PgClient) {
    await clearTestDb(db);
    await db.query(`CREATE TABLE test_data(
        id serial primary key,
        text text,
        uuid uuid,
        varchar varchar(255),
        bool bool,
        bytea bytea,
        int2 int2,
        int4 int4,
        int8 int8,
        float4 float4,
        float8 float8,
        numeric numeric, -- same as decimal
        json json,
        jsonb jsonb,
        pg_lsn pg_lsn,
        date date,
        time time,
        timestamp timestamp,
        timestamptz timestamptz,
        interval interval,
        macaddr macaddr,
        inet inet,
        oid oid
    )`);

    await db.query(`DROP TABLE IF EXISTS test_data_arrays`);
    await db.query(`CREATE TABLE test_data_arrays(
        id serial primary key,
        text text[],
        uuid uuid[],
        varchar varchar(255)[],
        bool bool[],
        bytea bytea[],
        int2 int2[],
        int4 int4[],
        int8 int8[],
        float4 float4[],
        float8 float8[],
        numeric numeric[], -- same as decimal
        json json[],
        jsonb jsonb[],
        pg_lsn pg_lsn[],
        date date[],
        time time[],
        timestamp timestamp[],
        timestamptz timestamptz[],
        interval interval[],
        macaddr macaddr[],
        inet inet[],
        oid oid[],
        multidimensional text[][] -- same as text[]
    )`);
  }

  async function insert(db: pgwire.PgClient) {
    await db.query(`
INSERT INTO test_data(id, text, uuid, varchar, bool, bytea, int2, int4, int8, numeric, float4, float8)
VALUES(1, 'text', 'baeb2514-4c57-436d-b3cc-c1256211656d', 'varchar', true, 'test', 1000, 1000000, 9007199254740993, 18014398509481982123, 3.14, 314);
    
INSERT INTO test_data(id, json, jsonb)
VALUES(2, '{"test": "thing" }', '{"test": "thing" }');

INSERT INTO test_data(id, date, time, timestamp, timestamptz)
VALUES(3, '2023-03-06', '15:47', '2023-03-06 15:47', '2023-03-06 15:47+02');

INSERT INTO test_data(id, pg_lsn, interval, macaddr, inet, oid)
VALUES(4, '016/B374D848', '1 hour', '00:00:5e:00:53:af', '127.0.0.1', 1007);

INSERT INTO test_data(id, date, time, timestamp, timestamptz)
VALUES(5, '-infinity'::date, 'allballs'::time, '-infinity'::timestamp, '-infinity'::timestamptz);

INSERT INTO test_data(id, timestamp, timestamptz)
VALUES(6, 'epoch'::timestamp, 'epoch'::timestamptz);

INSERT INTO test_data(id, timestamp, timestamptz)
VALUES(7, 'infinity'::timestamp, 'infinity'::timestamptz);

INSERT INTO test_data(id, timestamptz)
VALUES(8, '0022-02-03 12:13:14+03'::timestamptz);

INSERT INTO test_data(id, timestamptz)
VALUES(9, '10022-02-03 12:13:14+03'::timestamptz);
    `);
  }

  async function insertArrays(db: pgwire.PgClient) {
    await db.query(`
INSERT INTO test_data_arrays(id, text, uuid, varchar, bool, bytea, int2, int4, int8, numeric)
VALUES(1, ARRAY['text', '}te][xt{"'], '{"baeb2514-4c57-436d-b3cc-c1256211656d"}', '{"varchar"}', '{true}', '{"test"}', '{1000}', '{1000000}', '{9007199254740993}', '{18014398509481982123}');

INSERT INTO test_data_arrays(id, json, jsonb)
VALUES(2, ARRAY['{"test": "thing"}' :: json, '{"test": "}te][xt{"}' :: json], ARRAY['{"test": "thing", "foo": 5.0, "bignum": 18014398509481982123, "bool":true}' :: jsonb]);

INSERT INTO test_data_arrays(id, date, time, timestamp, timestamptz)
VALUES(3, ARRAY['2023-03-06'::date], ARRAY['15:47'::time], ARRAY['2023-03-06 15:47'::timestamp], ARRAY['2023-03-06 15:47+02'::timestamptz, '2023-03-06 15:47:00.123450+02'::timestamptz]);

INSERT INTO test_data_arrays(id, pg_lsn, interval, macaddr, inet, oid)
VALUES(4, ARRAY['016/B374D848'::pg_lsn], ARRAY['1 hour'::interval], ARRAY['00:00:5e:00:53:af'::macaddr], ARRAY['127.0.0.1'::inet], ARRAY[1007::oid]);

-- Empty arrays
INSERT INTO test_data_arrays(id, text, uuid, varchar, bool, bytea, int2, int4, int8, numeric)
VALUES(5, ARRAY[]::text[], ARRAY[]::uuid[], ARRAY[]::varchar[], ARRAY[]::bool[], ARRAY[]::bytea[], ARRAY[]::int2[], ARRAY[]::int4[], ARRAY[]::int8[], ARRAY[]::numeric[]);

-- Two-dimentional array
INSERT INTO test_data_arrays(id, multidimensional)
VALUES(6, ARRAY[['one', 1], ['two', 2], ['three', Null]]::TEXT[]);

-- Empty array
INSERT INTO test_data_arrays(id, multidimensional)
VALUES(7, ARRAY[[], [], []]::TEXT[]);

-- Empty array
INSERT INTO test_data_arrays(id, multidimensional)
VALUES(8, ARRAY[]::TEXT[]);

-- Array with only null
INSERT INTO test_data_arrays(id, multidimensional)
VALUES(9, ARRAY[NULL]::TEXT[]);

-- Array with 'null'
INSERT INTO test_data_arrays(id, multidimensional)
VALUES(10, ARRAY['null']::TEXT[]);
    `);
  }

  function checkResults(transformed: Record<string, any>[]) {
    expect(transformed[0]).toMatchObject({
      id: 1n,
      text: 'text',
      uuid: 'baeb2514-4c57-436d-b3cc-c1256211656d',
      varchar: 'varchar',
      bool: 1n,
      bytea: new Uint8Array([116, 101, 115, 116]),
      int2: 1000n,
      int4: 1000000n,
      int8: 9007199254740993n,
      float4: 3.14,
      float8: 314,
      numeric: '18014398509481982123'
    });
    expect(transformed[1]).toMatchObject({
      id: 2n,
      json: '{"test": "thing" }', // Whitespace preserved
      jsonb: '{"test": "thing"}' // Whitespace according to pg JSON conventions
    });

    expect(transformed[2]).toMatchObject({
      id: 3n,
      date: '2023-03-06',
      time: new TimeValue('15:47:00'),
      timestamp: new DateTimeValue('2023-03-06T15:47:00.000000', '2023-03-06 15:47:00'),
      timestamptz: new DateTimeValue('2023-03-06T13:47:00.000000Z', '2023-03-06 13:47:00Z')
    });

    expect(transformed[3]).toMatchObject({
      id: 4n,
      pg_lsn: '00000016/B374D848',
      interval: '01:00:00',
      macaddr: '00:00:5e:00:53:af',
      inet: '127.0.0.1',
      oid: 1007n
    });

    expect(transformed[4]).toMatchObject({
      id: 5n,
      date: '0000-01-01',
      time: new TimeValue('00:00:00'),
      timestamp: new DateTimeValue('0000-01-01T00:00:00'),
      timestamptz: new DateTimeValue('0000-01-01T00:00:00Z')
    });

    expect(transformed[5]).toMatchObject({
      id: 6n,
      timestamp: new DateTimeValue('1970-01-01T00:00:00.000000', '1970-01-01 00:00:00'),
      timestamptz: new DateTimeValue('1970-01-01T00:00:00.000000Z', '1970-01-01 00:00:00Z')
    });

    expect(transformed[6]).toMatchObject({
      id: 7n,
      timestamp: new DateTimeValue('9999-12-31T23:59:59'),
      timestamptz: new DateTimeValue('9999-12-31T23:59:59Z')
    });

    expect(transformed[7]).toMatchObject({
      id: 8n,
      timestamptz: new DateTimeValue('0022-02-03T09:13:14.000000Z', '0022-02-03 09:13:14Z')
    });

    expect(transformed[8]).toMatchObject({
      id: 9n,
      // 10022-02-03 12:13:14+03 - out of range of both our date parsing logic, and sqlite's date functions
      // We can consider just preserving the source string as an alternative if this causes issues.
      timestamptz: null
    });
  }

  function checkResultArrays(transformed: Record<string, any>[]) {
    expect(transformed[0]).toMatchObject({
      id: 1n,
      text: `["text","}te][xt{\\""]`,
      uuid: '["baeb2514-4c57-436d-b3cc-c1256211656d"]',
      varchar: '["varchar"]',
      bool: '[1]',
      bytea: '[null]',
      int2: '[1000]',
      int4: '[1000000]',
      int8: `[9007199254740993]`,
      numeric: `["18014398509481982123"]`
    });

    // Note: Depending on to what extent we use the original postgres value, the whitespace may change, and order may change.
    // We do expect that decimals and big numbers are preserved.
    expect(transformed[1]).toMatchObject({
      id: 2n,
      // Expected output after a serialize + parse cycle:
      // json: `[{"test":"thing"},{"test":"}te][xt{"}]`,
      // jsonb: `[{"foo":5.0,"bool":true,"test":"thing","bignum":18014398509481982123}]`
      // Expected using direct PG values:
      json: `[{"test": "thing"},{"test": "}te][xt{"}]`,
      jsonb: `[{"foo": 5.0, "bool": true, "test": "thing", "bignum": 18014398509481982123}]`
    });

    expect(transformed[2]).toMatchObject({
      id: 3n,
      date: `["2023-03-06"]`,
      time: `["15:47:00"]`,
      timestamp: '["2023-03-06 15:47:00"]',
      timestamptz: '["2023-03-06 13:47:00Z","2023-03-06 13:47:00.12345Z"]'
    });

    expect(transformed[3]).toMatchObject({
      id: 4n,
      pg_lsn: `["00000016/B374D848"]`,
      interval: `["01:00:00"]`,
      macaddr: `["00:00:5e:00:53:af"]`,
      inet: `["127.0.0.1"]`,
      oid: `[1007]`
    });

    expect(transformed[4]).toMatchObject({
      id: 5n,
      text: '[]',
      uuid: '[]',
      varchar: '[]',
      bool: '[]',
      bytea: '[]',
      int2: '[]',
      int4: '[]',
      int8: '[]',
      numeric: '[]'
    });

    expect(transformed[5]).toMatchObject({
      id: 6n,
      multidimensional: '[["one","1"],["two","2"],["three",null]]'
    });

    expect(transformed[6]).toMatchObject({
      id: 7n,
      multidimensional: '[]'
    });

    expect(transformed[7]).toMatchObject({
      id: 8n,
      multidimensional: '[]'
    });

    expect(transformed[8]).toMatchObject({
      id: 9n,
      multidimensional: '[null]'
    });

    expect(transformed[9]).toMatchObject({
      id: 10n,
      multidimensional: '["null"]'
    });
  }

  test('test direct queries', async () => {
    const db = await connectPgPool();
    try {
      await setupTable(db);

      await insert(db);

      const transformed = [
        ...WalStream.getQueryData(pgwire.pgwireRows(await db.query(`SELECT * FROM test_data ORDER BY id`)))
      ];

      checkResults(transformed);
    } finally {
      await db.end();
    }
  });

  test('test direct queries - parameterized', async () => {
    // Parameterized queries may use a different underlying protocol,
    // so we make sure it also gets the same results.
    const db = await connectPgPool();
    try {
      await setupTable(db);

      await insert(db);

      const transformed = [
        ...WalStream.getQueryData(
          pgwire.pgwireRows(
            await db.query({
              statement: `SELECT * FROM test_data WHERE $1 ORDER BY id`,
              params: [{ type: 'bool', value: true }]
            })
          )
        )
      ];

      checkResults(transformed);
    } finally {
      await db.end();
    }
  });

  test('test direct queries - arrays', async () => {
    const db = await connectPgPool();
    try {
      await setupTable(db);

      await insertArrays(db);

      const transformed = [
        ...WalStream.getQueryData(pgwire.pgwireRows(await db.query(`SELECT * FROM test_data_arrays ORDER BY id`)))
      ].map((e) => applyRowContext(e, CompatibilityContext.FULL_BACKWARDS_COMPATIBILITY));

      checkResultArrays(transformed);
    } finally {
      await db.end();
    }
  });

  test('test replication', async () => {
    const db = await connectPgPool();
    try {
      await setupTable(db);

      const slotName = 'test_slot';

      await db.query({
        statement: 'SELECT pg_drop_replication_slot(slot_name) FROM pg_replication_slots WHERE slot_name = $1',
        params: [{ type: 'varchar', value: slotName }]
      });

      await db.query({
        statement: `SELECT slot_name, lsn FROM pg_catalog.pg_create_logical_replication_slot($1, 'pgoutput')`,
        params: [{ type: 'varchar', value: slotName }]
      });

      await insert(db);

      const pg: pgwire.PgConnection = await pgwire.pgconnect({ replication: 'database' }, TEST_URI);
      const replicationStream = await pg.logicalReplication({
        slot: slotName,
        options: {
          proto_version: '1',
          publication_names: 'powersync'
        }
      });

      const transformed = await getReplicationTx(db, replicationStream);
      await pg.end();

      checkResults(transformed);
    } finally {
      await db.end();
    }
  });

  test('test replication - arrays', async () => {
    const db = await connectPgPool();
    try {
      await setupTable(db);

      const slotName = 'test_slot';

      await db.query({
        statement: 'SELECT pg_drop_replication_slot(slot_name) FROM pg_replication_slots WHERE slot_name = $1',
        params: [{ type: 'varchar', value: slotName }]
      });

      await db.query({
        statement: `SELECT slot_name, lsn FROM pg_catalog.pg_create_logical_replication_slot($1, 'pgoutput')`,
        params: [{ type: 'varchar', value: slotName }]
      });

      await insertArrays(db);

      const pg: pgwire.PgConnection = await pgwire.pgconnect({ replication: 'database' }, TEST_URI);
      const replicationStream = await pg.logicalReplication({
        slot: slotName,
        options: {
          proto_version: '1',
          publication_names: 'powersync'
        }
      });

      const transformed = await getReplicationTx(db, replicationStream);
      await pg.end();

      checkResultArrays(transformed.map((e) => applyRowContext(e, CompatibilityContext.FULL_BACKWARDS_COMPATIBILITY)));
    } finally {
      await db.end();
    }
  });

  test('schema', async function () {
    const db = await connectPgWire();

    await setupTable(db);

    // TODO need a test for adapter
    // const schema = await api.getConnectionsSchema(db);
    // expect(schema).toMatchSnapshot();
  });

  test('date formats', async () => {
    const db = await connectPgWire();
    try {
      await setupTable(db);

      await db.query(`
INSERT INTO test_data(id, time, timestamp, timestamptz) VALUES (1, '17:42:01.12', '2023-03-06 15:47:12.4', '2023-03-06 15:47+02');
`);

      const [row] = [
        ...WalStream.getQueryData(
          pgwire.pgwireRows(await db.query(`SELECT time, timestamp, timestamptz FROM test_data`))
        )
      ];

      const oldFormat = applyRowContext(row, CompatibilityContext.FULL_BACKWARDS_COMPATIBILITY);
      expect(oldFormat).toMatchObject({
        time: '17:42:01.12',
        timestamp: '2023-03-06 15:47:12.4',
        timestamptz: '2023-03-06 13:47:00Z'
      });

      const newFormat = applyRowContext(row, new CompatibilityContext(CompatibilityEdition.SYNC_STREAMS));
      expect(newFormat).toMatchObject({
        time: '17:42:01.120000',
        timestamp: '2023-03-06T15:47:12.400000',
        timestamptz: '2023-03-06T13:47:00.000000Z'
      });
    } finally {
      await db.end();
    }
  });

  test('test replication - custom types', async () => {
    const db = await connectPgPool();
    try {
      await clearTestDb(db);
      await db.query(`CREATE DOMAIN rating_value AS FLOAT CHECK (VALUE BETWEEN 0 AND 5);`);
      await db.query(`CREATE TYPE composite AS (foo rating_value, bar TEXT);`);

      await db.query(`CREATE TABLE test_custom(
        id serial primary key,
        rating rating_value,
        composite composite
      );`);

      const slotName = 'test_slot';

      await db.query({
        statement: 'SELECT pg_drop_replication_slot(slot_name) FROM pg_replication_slots WHERE slot_name = $1',
        params: [{ type: 'varchar', value: slotName }]
      });

      await db.query({
        statement: `SELECT slot_name, lsn FROM pg_catalog.pg_create_logical_replication_slot($1, 'pgoutput')`,
        params: [{ type: 'varchar', value: slotName }]
      });

      await db.query(`
        INSERT INTO test_custom
          (rating, composite)
        VALUES (
          1,
          (2, 'bar')
        );
      `);

      const pg: pgwire.PgConnection = await pgwire.pgconnect({ replication: 'database' }, TEST_URI);
      const replicationStream = await pg.logicalReplication({
        slot: slotName,
        options: {
          proto_version: '1',
          publication_names: 'powersync'
        }
      });

      const [transformed] = await getReplicationTx(db, replicationStream);
      await pg.end();

      const oldFormat = applyRowContext(transformed, CompatibilityContext.FULL_BACKWARDS_COMPATIBILITY);
      expect(oldFormat).toMatchObject({
        rating: '1'
      });

      const newFormat = applyRowContext(transformed, new CompatibilityContext(CompatibilityEdition.SYNC_STREAMS));
      expect(newFormat).toMatchObject({
        rating: 1
      });
    } finally {
      await db.end();
    }
  });
});

/**
 * Return all the inserts from the first transaction in the replication stream.
 */
async function getReplicationTx(db: pgwire.PgClient, replicationStream: pgwire.ReplicationStream) {
  const typeCache = new PostgresTypeCache(db);
  await typeCache.fetchTypesForSchema();

  let transformed: SqliteInputRow[] = [];
  for await (const batch of replicationStream.pgoutputDecode()) {
    for (const msg of batch.messages) {
      if (msg.tag == 'insert') {
        transformed.push(typeCache.constructAfterRecord(msg));
      } else if (msg.tag == 'commit') {
        return transformed;
      }
    }
  }
  return transformed;
}
