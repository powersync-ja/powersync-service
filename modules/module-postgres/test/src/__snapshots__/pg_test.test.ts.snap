// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`pg data types > schema 1`] = `
[
  {
    "name": "public",
    "tables": [
      {
        "columns": [
          {
            "name": "id",
            "pg_type": "int4",
            "type": "integer",
          },
          {
            "name": "text",
            "pg_type": "text",
            "type": "text",
          },
          {
            "name": "uuid",
            "pg_type": "uuid",
            "type": "uuid",
          },
          {
            "name": "varchar",
            "pg_type": "varchar",
            "type": "character varying(255)",
          },
          {
            "name": "bool",
            "pg_type": "bool",
            "type": "boolean",
          },
          {
            "name": "bytea",
            "pg_type": "bytea",
            "type": "bytea",
          },
          {
            "name": "int2",
            "pg_type": "int2",
            "type": "smallint",
          },
          {
            "name": "int4",
            "pg_type": "int4",
            "type": "integer",
          },
          {
            "name": "int8",
            "pg_type": "int8",
            "type": "bigint",
          },
          {
            "name": "float4",
            "pg_type": "float4",
            "type": "real",
          },
          {
            "name": "float8",
            "pg_type": "float8",
            "type": "double precision",
          },
          {
            "name": "numeric",
            "pg_type": "numeric",
            "type": "numeric",
          },
          {
            "name": "json",
            "pg_type": "json",
            "type": "json",
          },
          {
            "name": "jsonb",
            "pg_type": "jsonb",
            "type": "jsonb",
          },
          {
            "name": "pg_lsn",
            "pg_type": "pg_lsn",
            "type": "pg_lsn",
          },
          {
            "name": "date",
            "pg_type": "date",
            "type": "date",
          },
          {
            "name": "time",
            "pg_type": "time",
            "type": "time without time zone",
          },
          {
            "name": "timestamp",
            "pg_type": "timestamp",
            "type": "timestamp without time zone",
          },
          {
            "name": "timestamptz",
            "pg_type": "timestamptz",
            "type": "timestamp with time zone",
          },
          {
            "name": "interval",
            "pg_type": "interval",
            "type": "interval",
          },
          {
            "name": "macaddr",
            "pg_type": "macaddr",
            "type": "macaddr",
          },
          {
            "name": "inet",
            "pg_type": "inet",
            "type": "inet",
          },
          {
            "name": "oid",
            "pg_type": "oid",
            "type": "oid",
          },
        ],
        "name": "test_data",
      },
      {
        "columns": [
          {
            "name": "id",
            "pg_type": "int4",
            "type": "integer",
          },
          {
            "name": "text",
            "pg_type": "text[]",
            "type": "text[]",
          },
          {
            "name": "uuid",
            "pg_type": "uuid[]",
            "type": "uuid[]",
          },
          {
            "name": "varchar",
            "pg_type": "varchar[]",
            "type": "character varying(255)[]",
          },
          {
            "name": "bool",
            "pg_type": "bool[]",
            "type": "boolean[]",
          },
          {
            "name": "bytea",
            "pg_type": "bytea[]",
            "type": "bytea[]",
          },
          {
            "name": "int2",
            "pg_type": "int2[]",
            "type": "smallint[]",
          },
          {
            "name": "int4",
            "pg_type": "int4[]",
            "type": "integer[]",
          },
          {
            "name": "int8",
            "pg_type": "int8[]",
            "type": "bigint[]",
          },
          {
            "name": "float4",
            "pg_type": "float4[]",
            "type": "real[]",
          },
          {
            "name": "float8",
            "pg_type": "float8[]",
            "type": "double precision[]",
          },
          {
            "name": "numeric",
            "pg_type": "numeric[]",
            "type": "numeric[]",
          },
          {
            "name": "json",
            "pg_type": "json[]",
            "type": "json[]",
          },
          {
            "name": "jsonb",
            "pg_type": "jsonb[]",
            "type": "jsonb[]",
          },
          {
            "name": "pg_lsn",
            "pg_type": "pg_lsn[]",
            "type": "pg_lsn[]",
          },
          {
            "name": "date",
            "pg_type": "date[]",
            "type": "date[]",
          },
          {
            "name": "time",
            "pg_type": "time[]",
            "type": "time without time zone[]",
          },
          {
            "name": "timestamp",
            "pg_type": "timestamp[]",
            "type": "timestamp without time zone[]",
          },
          {
            "name": "timestamptz",
            "pg_type": "timestamptz[]",
            "type": "timestamp with time zone[]",
          },
          {
            "name": "interval",
            "pg_type": "interval[]",
            "type": "interval[]",
          },
          {
            "name": "macaddr",
            "pg_type": "macaddr[]",
            "type": "macaddr[]",
          },
          {
            "name": "inet",
            "pg_type": "inet[]",
            "type": "inet[]",
          },
          {
            "name": "oid",
            "pg_type": "oid[]",
            "type": "oid[]",
          },
          {
            "name": "multidimensional",
            "pg_type": "text[]",
            "type": "text[]",
          },
        ],
        "name": "test_data_arrays",
      },
    ],
  },
]
`;
